drivers: modem: add support for gemalto ehs6

cybertale opened the pull request:
This driver supports TCP/UDP client on ehs6 module, which supports IPV4
internet connection with AT commands. This driver has been tested on a
frdm_k64f board.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

mike-scott commented:
@cybertale Thank you for your submission.   As it is the weekend and I have a holiday on Monday, I would ask for your patience as I get time to do a review.

cybertale commented:
> @cybertale Thank you for your submission. As it is the weekend and I have a holiday on Monday, I would ask for your patience as I get time to do a review.

Sure, and here is the AT command manual of this module. 
https://m2m.dk/wp-content/uploads/2014/05/EHS6-AT-Command-Set-V03.001-19.02.2015-1.pdf

mike-scott commented on drivers/modem/Kconfig.ehs6:
```diff
@@ -0,0 +1,65 @@
+	select NET_OFFLOAD
+	select UART_MCUX_2 if BOARD_FRDM_K64F
+	help
+	  Choose this setting to enable Wistron WNC-M14A2A LTE-M modem driver.
```
Copy/pasta mentioning Wistron modem

cybertale replied:
Sorry, awkward...

mike-scott commented on drivers/modem/Kconfig.ehs6:
```diff
@@ -0,0 +1,65 @@
+	select UART_MCUX_2 if BOARD_FRDM_K64F
+	help
+	  Choose this setting to enable Wistron WNC-M14A2A LTE-M modem driver.
+	  NOTE: Currently the pin settings only work with FRDM K64F shield.
```
Is this true for this modem?  Or is this copy/paste issue as well?

cybertale replied:
Sorry, I copied the Wistron driver but forgot to modify the help of Kconfig file.

mike-scott commented on drivers/modem/Kconfig.ehs6:
```diff
@@ -0,0 +1,65 @@
+if MODEM_EHS6
+
+config MODEM_EHS6_RX_STACK_SIZE
+	int "Size of the stack for the WNC-M14A2A modem driver RX thread"
```
WNC-14A2A mention

mike-scott commented on drivers/modem/Kconfig.ehs6:
```diff
@@ -0,0 +1,65 @@
+	help
+	  This setting is used in the AT%PDNSET command to set the APN name
+	  for the network connection context.  Normally, don't need to change
+	  this value.
```
I mentioned "Normally, don't need to change this value" in the Wistron modem settings, because it's ordered through the AT&T store with an AT&T IoT SIM.  This may not be true for your modem.

mike-scott commented on drivers/modem/Kconfig.ehs6:
```diff
@@ -0,0 +1,65 @@
+	  this value.
+
+config MODEM_EHS6_INIT_PRIORITY
+	int "WNC-M14A2A driver init priority"
```
WNC-M14A2A mention

mike-scott commented on drivers/modem/Kconfig.ehs6:
```diff
@@ -0,0 +1,65 @@
+	depends on MODEM_EHS6
+	default 80
+	help
+	  WNC-M14A2A device driver initialization priority.
```
WNC-M14A2A mention

mike-scott commented on drivers/modem/Kconfig.ehs6:
```diff
@@ -0,0 +1,65 @@
+	  so that it can start before the networking sub-system.
+
+config MODEM_EHS6_TIMEOUT_TIME
+	string "WNC-M14A2A driver init priority"
```
WNC-M14A2A mention

mike-scott commented on drivers/modem/Kconfig.ehs6:
```diff
@@ -0,0 +1,65 @@
+	depends on MODEM_EHS6
+	default "80"
+	help
+	  WNC-M14A2A device driver initialization priority.
```
WNC-M14A2A mention

mike-scott commented on drivers/modem/ehs6.c:
```diff
@@ -0,0 +1,1005 @@
+ */
+/* #define ENABLE_VERBOSE_MODEM_RECV_HEXDUMP	1 */
+
+#define MDM_UART_DEV_NAME		DT_UART_MCUX_2_NAME
```
This is currently hard-coded for only the K64F UART 2, is that correct?  Seems like this could be made generic the same was Wistron modem does it by placing the modem node inside the uart device it's attached to.   See:
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/modem/wncm14a2a.c#L94

The `DT_WNC_M14A2A_0_BUS_NAME` define is dynamically generated via DTS based on which UART node the WNC modem DTS is placed inside.

mike-scott commented on drivers/modem/Kconfig.ehs6:
```diff
@@ -0,0 +1,65 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+#CONFIG_OFFLOAD_SOCKET
```
Does this need to be removed?

mike-scott commented on drivers/modem/Kconfig.ehs6:
```diff
@@ -0,0 +1,65 @@
+	bool "Enable Gemalto EHS6 modem driver"
+	depends on MODEM
+	select MODEM_RECEIVER
+	select NET_OFFLOAD
```
I believe you'll also want to also select  `NET_SOCKETS_OFFLOAD` as you're using those APIs as well as the `NET_OFFLOAD` APIs.

cybertale replied:
That's right.

mike-scott reviewed (changes requested):
Found some copy/paste issues, but I like that you used NET_SOCKETS_OFFLOAD APIs.  (I will be moving the Wistron modem to those APIs as well.)   I will take some time to read the docs on your modem and provide a more detailed review once you've updated the PR.

mike-scott commented:
@cybertale I realize now looking at the PR again, that this does not support DTS at the moment.  Did you plan to update that in this PR or add it as a second PR later?

cybertale commented:
@mike-scott DTS support is not a big problem. What I'm worried about is that I have about nearly about 500 lines of code copied from your driver, which mainly is about pulling data from the uart (ehs6_rx() I mean). We should make this generic but I don't know many modems and don't know if all modems use UART and AT commands. What do you think we should do with this?

mike-scott commented:
> @mike-scott DTS support is not a big problem. What I'm worried about is that I have about nearly about 500 lines of code copied from your driver, which mainly is about pulling data from the uart (ehs6_rx() I mean). We should make this generic but I don't know many modems and don't know if all modems use UART and AT commands. What do you think we should do with this?

@cybertale Yes, I'm researching the best way to implement a UART/AT command (3GPP?) API for modems to use so that we don't copy that code.

cybertale commented:
@mike-scott This is good, how long do you think would it take? My PR can wait for that change while I keep fixing problems about it in the same time. Or merge this then make that change?

mike-scott commented:
@cybertale The earliest this PR could be merged would be in the 1.15 window.  I've been very busy with work so I haven't had the time to see what other RTOS are doing in this area, etc.

mike-scott commented:
FYI, I also have another modem driver waiting in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/14057  And I would try and solve all 3 modems with the new UART/AT command API.

cybertale committed:
drivers: modem: add support for gemalto ehs6

This driver supports TCP/UDP client on ehs6 module, which supports IPV4
internet connection with AT commands. This driver has been tested on a
frdm_k64f board.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale commented:
@mike-scott Hi, sorry for the late, I added DTS support and fixed copy/paste issues. Give it a look when you got some time.

mike-scott commented:
@cybertale Sorry for the delay.  I had a rather large modem PR which was just merged, so I can now consider a new PR which separates generic handling of the UART into common code.

We have the 2 modem drivers in mainline Zephyr, plus this PR and another driver which is not public.  They all have different requirements and it will take me some time to come up with a suitable solution to present.

But at least it's now in-progress.

cybertale commented:
@mike-scott Sure, take your time.

ioannisg commented:
@cybertale @mike-scott  is it ok to move this to 2.1 release?

galak commented:
Any updates on this?
